### PR TITLE
Adding initial type support related tests for BQ

### PR DESCRIPTION
**Pull request by adchia**
**What this PR does / why we need it**:
Adds tests that tests given input data of different types (for entities vs features), whether the resultant feature vectors from get_historical_features and get_online_features and inference match expectations.

As part of this, updating inference to respect entity join_key.

**Which issue(s) this PR fixes**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: sdk/python/feast/inference.py [hunk 1]
```diff
@@ -29,9 +29,17 @@ def update_entities_with_inferred_types_from_feature_views(
         col_names_and_types = view.batch_source.get_table_column_names_and_types(config)
         for entity_name in view.entities:
             if entity_name in incomplete_entities:
+                entity = incomplete_entities[entity_name]
+                entity_table_col_join_key = (
+                    entity.join_key if entity.join_key else entity.name
```
Comment (felixwang9817) [inline sdk/python/feast/inference.py:34]: Why not just `entity.join_key`? If `join_key` is not specified in the initialization of an Entity, it will be set to `name`.

File: sdk/python/feast/inference.py [hunk 2]
```diff
@@ -29,9 +29,17 @@ def update_entities_with_inferred_types_from_feature_views(
         col_names_and_types = view.batch_source.get_table_column_names_and_types(config)
         for entity_name in view.entities:
             if entity_name in incomplete_entities:
+                entity = incomplete_entities[entity_name]
+                entity_table_col_join_key = (
```
Comment (felixwang9817) [inline sdk/python/feast/inference.py:33]: Nit: the name `entity_table_col_join_key` seems a bit verbose. What about just `entity_join_key`?

File: sdk/python/tests/integration/registration/test_types.py [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+    }
+    non_list_val = dtype_map[dtype]
+    # Duplicate the value once if this is a list
+    if is_list:
+        return [[n, n] if n is not None else None for n in non_list_val]
+    else:
+        return non_list_val
+
+
+def create_dataset(
```
Comment (felixwang9817) [inline sdk/python/tests/integration/registration/test_types.py:54]: I think all the dataset creation functions can go in `tests/data/data_creator.py`. At least, that's what the structure of the first universal feature repo PR (#1734) suggests.

File: sdk/python/feast/inference.py (hunk 2 above)
Comment (adchia) [reply sdk/python/feast/inference.py]: Done

File: sdk/python/feast/inference.py (hunk 1 above)
Comment (adchia) [reply sdk/python/feast/inference.py]: Good call. Changed

File: sdk/python/tests/integration/registration/test_types.py (hunk 3 above)
Comment (adchia) [reply sdk/python/tests/integration/registration/test_types.py]: Moved to universal repo. PTAL!

File: sdk/python/tests/integration/e2e/test_universal_e2e.py [hunk 4]
```diff
@@ -78,7 +78,7 @@ def check_offline_and_online_features(
 def run_offline_online_store_consistency_test(
     fs: FeatureStore, fv: FeatureView
 ) -> None:
-    now = datetime.utcnow()
+    now = datetime.now()
```
Comment (achals) [inline sdk/python/tests/integration/e2e/test_universal_e2e.py:81]: Why was this changed to `now` from `utcnow`?

File: sdk/python/tests/integration/feature_repos/test_repo_configuration.py [hunk 5]
```diff
@@ -208,9 +214,14 @@ def construct_test_environment(
...
     :param test_repo_config: configuration
+    :param create_and_apply: whether to create and apply the repo config
+    :param materialize: whether to materialize features to online store
+    :param entity_type: the data type for the entity column (i.e. id)
+    :param feature_dtype: the data type for the feature column (i.e. value)
+    :param feature_is_list: whether the feature column (i.e. value) should be a list feature
     :return: A feature store built using the supplied configuration.
     """
-    df = create_dataset()
+    df = create_dataset(entity_type, feature_dtype, feature_is_list)
```
Comment (achals) [inline sdk/python/tests/integration/feature_repos/test_repo_configuration.py:224]: I think this is cool for now, but just a note that I'm refactoring this path a bit so the dataset creation is more pluggable to allow for custom datasets a little more easily.

File: sdk/python/tests/integration/feature_repos/universal/entities.py [hunk 6]
```diff
@@ -10,5 +10,10 @@ def driver():
     )
 
 
+def driver_no_value_type():
+    # Don't specify value type in entity to force inference
+    return Entity(name="driver", description="driver id", join_key="driver_id",)
+
+
```
Comment (achals) [inline sdk/python/tests/integration/feature_repos/universal/entities.py:17]: Do you want to use the existing driver entity but add some params to customize it? Like:
```
def driver(value=True):
    return Entity(
        name="driver",  # The name is derived from this argument, not object name.
        value_type=ValueType.INT64 if Value else ValueType.UNKNOWN,
        description="driver id",
        join_key="driver_id",
    )
```

File: sdk/python/tests/integration/materialization/test_offline_online_store_consistency.py [hunk 7]
```diff
@@ -332,7 +332,7 @@ def check_offline_and_online_features(
 def run_offline_online_store_consistency_test(
     fs: FeatureStore, fv: FeatureView, full_feature_names: bool,
 ) -> None:
-    now = datetime.utcnow()
+    now = datetime.now()
```
Comment (achals) [inline sdk/python/tests/integration/materialization/test_offline_online_store_consistency.py:335]: same q

**Review by achals — APPROVED** [review]
looks good with some nits.

File: sdk/python/tests/integration/e2e/test_universal_e2e.py (hunk 4 above)
Comment (adchia) [reply sdk/python/tests/integration/e2e/test_universal_e2e.py:81]: We have some places where we used utcnow and other places where we used now, which caused cases where we sometimes got no results when doing a historical retrieval

File: sdk/python/tests/integration/feature_repos/universal/entities.py (hunk 6 above)
Comment (adchia) [reply sdk/python/tests/integration/feature_repos/universal/entities.py:17]: Done!

File: sdk/python/tests/integration/materialization/test_offline_online_store_consistency.py (hunk 7 above)
Comment (adchia) [reply sdk/python/tests/integration/materialization/test_offline_online_store_consistency.py:335]: same response. create_data_set was always using utcnow, but there are other tests that don't use utcnow and instead use now(). This unifies across our test suites

File: sdk/python/tests/integration/feature_repos/test_repo_configuration.py (hunk 5 above)
Comment (adchia) [reply sdk/python/tests/integration/feature_repos/test_repo_configuration.py:224]: Noted. Would you rather I wait to merge this in until after your refactor then?

**Review by achals — APPROVED** [review]
/lgtm
